Refactor and fix all upgrade integration tests

blakerouse opened the pull request:
<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement
- Breaking change
- Deprecation
- Cleanup
- Docs
-->

## What does this PR do?

<!-- Mandatory
Explain here the changes you made on the PR. Please explain the WHAT: patterns used, algorithms implemented, design architecture, message processing, etc.
-->

This refactors and fixes all upgrade integration tests. This provides a unified path for the tests to follow for testing upgrade. Includes code to make the upgrades faster, and allows downgrades to work. Includes fixes for upgrade tests that where skipped because they where flaky and enables the retry upgrade test on Windows.

## Why is it important?

<!-- Mandatory
Explain here the WHY, or the rationale/motivation for the changes.
-->

Ensures that all upgrade tests are working properly so we can catch real errors instead of fighting the CI.

## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- ~~[ ] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)~~
- [x] I have added an integration test or an E2E test

## How to test this PR locally

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

`mage integration:test`

## Related issues

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Relates #123
- Requires #123
- Superseds #123
-->
- Closes #3155
- Closes #3461

blakerouse commented:
This has a replace present until this PR is merged and new elastic-agent-libs is released - https://github.com/elastic/elastic-agent-libs/pull/152

elasticmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent%2Felastic-agent-mbp%2FPR-3477/detail/PR-3477/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent%2Felastic-agent-mbp%2FPR-3477/detail/PR-3477/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent%2Felastic-agent-mbp%2FPR-3477/detail/PR-3477/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent%2Felastic-agent-mbp%2FPR-3477/detail/PR-3477/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://elastic-agent_3477.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/fleet-ci/transactions/view?rangeFrom=2023-09-26T17:31:34.073Z&rangeTo=2023-09-26T17:51:34.073Z&transactionName=BUILD+elastic-agent%2Felastic-agent-mbp%2FPR-%7Bnumber%7D&transactionType=job&latencyAggregationType=avg&traceId=6d55f4d93149fa4f36f7085dc795db70&transactionId=9042d5607265d0b6)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2023-09-26T17:41:34.073+0000


* Duration: 6 min 29 sec





</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->














### Steps errors [![1](https://img.shields.io/badge/1%20-red)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent%2Felastic-agent-mbp%2FPR-3477/detail/PR-3477/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `check`
<ul>
<li>Took 1 min 7 sec . View more details <a href="https://fleet-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/elastic-agent/pipelines/elastic-agent-mbp/pipelines/PR-3477/runs/1/steps/175/log/?start=0">here</a></li>
<li>Description: <code>make check-ci</code></l1>
</ul>
  
  </p>
  </details>



## :grey_exclamation: Flaky test report
No test was executed to be analysed.



## :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.


- `/package` : Generate the packages.


- `run integration tests` : Run the Elastic Agent Integration tests.


- `run end-to-end tests` : Generate the packages and run the E2E Tests.


- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)




</p>
</details>

<!--COMMENT_GENERATED_WITH_ID_comment.id-->

elasticmachine commented:
### :globe_with_meridians: Coverage report

Name | Metrics % (`covered/total`) | Diff
 --- | --- | ---
Packages | 98.78% (`81/82`) | :+1: 
Files | 66.102% (`195/295`) | :+1: 
Classes | 65.574% (`360/549`) | :-1: -0.12
Methods | 52.695% (`1134/2152`) | :-1: -0.049
Lines | 38.131% (`12862/33731`) | :-1: -0.039
Conditionals | 100.0% (`0/0`) | :green_heart: 

<!--COMMENT_GENERATED_WITH_ID_coverage-->

cmacknz commented:
`testing/**` needs to be added to the SonarQube exclusion list to silence the coverage failure.

https://github.com/elastic/elastic-agent/blob/main/sonar-project.properties

cmacknz commented on pkg/testing/fixture.go:
```diff
@@ -456,6 +456,23 @@ func (f *Fixture) ExecInspect(ctx context.Context, opts ...process.CmdOption) (A
 	return inspect, err
 }
 
+// ExecVersion executes to version subcommand on the prepared Elastic Agent binary
+// with '--binary-only'. It returns the parsed YAML output.
```
```suggestion
// ExecVersion executes the version subcommand on the prepared Elastic Agent binary
// with '--binary-only'. It returns the parsed YAML output.
```

cmacknz commented on testing/upgradetest/upgrader.go:
```diff
@@ -0,0 +1,361 @@
+	logger.Logf("Upgrading from version %q to version %q", startParsedVersion, endVersionInfo.Binary.String())
+
+	// always pull from local path (as fixture is required)
+	// doesn't mean
```
> // doesn't mean

Doesn't mean what?

blakerouse replied:
both comments are wrong, will remove it

cmacknz commented on testing/upgradetest/watcher.go:
```diff
@@ -0,0 +1,106 @@
+// ConfigureFastWatcher writes an Elastic Agent configuration that
+// adjusts the watcher to be faster.
+func ConfigureFastWatcher(ctx context.Context, f *atesting.Fixture) error {
+	return f.Configure(ctx, []byte(FastWatcherCfg))
```
Does this work if the agent is Fleet managed?

cmacknz replied:
I think from previous testing it doesn't, since the only config the agent respects is the one from Fleet. We should note this here to avoid people using this where it doesn't help (unless I'm wrong about it).

blakerouse replied:
I think your correct, I will add a note.

cmacknz commented on testing/upgradetest/watcher.go:
```diff
@@ -0,0 +1,106 @@
+			for _, pid := range pids {
+				proc, err := os.FindProcess(pid)
+				if err == nil {
+					_ = proc.Kill()
```
Should we fail the test if we need to do this in a version of the agent that has the watcher killed as part of the uninstall?

blakerouse replied:
I will add a specific test that checks for that case. I prefer to keep this here always just to ensure that it doesn't cause other unrelated tests to break. That way its clear why that one test fails.

blakerouse replied:
I added `TestStandaloneUpgradeUninstallKillWatcher` for testing this specific behavior.

blakerouse commented on pkg/testing/fetcher_artifact.go:
```diff
@@ -68,7 +68,9 @@ func (f *artifactFetcher) Fetch(ctx context.Context, operatingSystem string, arc
 	preVersion := version
 	if uri == "" {
-		version = fmt.Sprintf("%s-SNAPSHOT", version)
+		if !strings.HasSuffix(version, "-SNAPSHOT") {
+			version += "-SNAPSHOT"
+		}
```
We actually did force it there. What didn't work was when you passed in a version string that had the suffix `-SNAPSHOT` it would append it again, which resulted in `-SNAPSHOT-SNAPSHOT` which was incorrect.

pchila commented on testing/integration/upgrade_broken_package_test.go:
```diff
@@ -0,0 +1,92 @@
+	require.NoError(t, err)
+
+	// Upgrade to an old build, see `BackwardTwoMinors` for why.
+	upgradeToVersion, err := upgradetest.BackwardTwoMinors(define.Version())
```
Why do we use BackwardTwoMinors here? 
We already know that GetPreviousMinor() makes assumptions about the version numbers (set the patch version to 0, return a fixed version for 7.x), why not taking the version from artifact API as it was before the refactor?

blakerouse replied:
Almost all the tests where written completely different, but basically doing the same thing. I tried my best to merge them onto a path that is very similar and easy to follow. To me read ability is the most important thing, because if I cannot open the test and easily read and understand what is going on then its useless.

I tried my best to align the tests in a common pattern. If I have change the meaning of what the test is testing please let me know I will fix it.

Does this change the meaning of this test?

For why we use `BackwardTwoMinors` check the docstring on that function. It explains it.

pchila replied:
It doesn't change the meaning of the test, I am not that comfortable with us not using the latest patch for a given <major>.<minor>

blakerouse replied:
Are you saying that you want it to only go down a single version? I could possibly switch all to that way, and then possibly only go down 2 in the case the last 1 doesn't exist.

pchila replied:
Note: this is not blocking for the merge

It's not the number of version we downgrade to: in this specific case it doesn't really matter how far back we go, we just want to test that even with a broken package version we can upgrade to something else.

The thing making me uncomfortable is that from `8.11.0-SNAPSHOT` we are going back to `8.10.0` or `8.9.0` instead of `8.10.3` or `8.9.2` because we cannot guess how patch versions a certain major.minor combination has.

As a personal preference I favored using [artifact api version list](https://artifacts-api.elastic.co/v1/versions) to search for a previous version instead of computing a likely value but it's true that unless we have serious bugs on our `x.y.0` versions this won't break any time soon for 8.x

7.x may break when 7.17.10 will no longer be available through artifact api (have a look at GetPreviousMinor() code)
9.x if and when it will land will need also some special handling in GetPreviousMinor()

As long as we use this to go a couple of minors  back from the latest 8.x we will not fall into those cases so it comes down to preference.

blakerouse replied:
I have updated the code to take the same idea you had from your other PR and use it in this PR. Now it uses the artifacts API to find the previous version that has a snapshot build and uses that for the test.

pchila commented on testing/integration/upgrade_fleet_test.go:
```diff
@@ -0,0 +1,179 @@
+	require.NoError(t, err)
+	endVersionInfo, err := endFixture.ExecVersion(ctx)
+	require.NoError(t, err)
+	if startVersionInfo.Binary.Commit == endVersionInfo.Binary.Commit {
+		t.Skipf("Build under test is the same as the build from the artifacts repository (commit: %s)", startVersionInfo.Binary.Commit)
+	}
```
Heads up: this does not work for repackaged agents (different agent package versions will have the same commit for the agent binary but different components).
For the moment, it will work

blakerouse replied:
How would that ever be the case for a PR? A PR should always result in a new build, and as a developer testing from your laptop it should always be from a custom build as well.

I don't see a reality where that is not true.

pchila replied:
If this is only for PRs run on CI it will work (because the merge commit is always different from whatever we may download from Artifact API)
In the case of a dev testing, it can happen once https://github.com/elastic/elastic-agent/issues/2579 and https://github.com/elastic/elastic-agent/issues/2780 are implemented (somebody is packaging uncommitted stuff on top of main and use a different AGENT_PACKAGE_VERSION value)

I agree it's not an immediate concern and the local case can be solved by a simple commit on whatever branch

blakerouse replied:
What about checking both the versions and the commit in the same compare, would that fix the compare when AGENT_PACKAGE_VERSION is set?

pchila replied:
Yep that would work nicely :+1:

pchila commented on testing/integration/upgrade_fleet_test.go:
```diff
@@ -0,0 +1,179 @@
+	t.Logf("Testing Elastic Agent upgrade from %s to %s with Fleet...", define.Version(), endVersionInfo.Binary.String())
+
+	// start fixture gets the agent configured to use a faster watcher
+	err = upgradetest.ConfigureFastWatcher(ctx, startFixture)
+	require.NoError(t, err)
```
Does this work for Fleet managed agents now?

blakerouse replied:
No it doesn't. I will remove this line.

pchila commented on testing/integration/upgrade_downgrade_test.go:
```diff
@@ -0,0 +1,49 @@
+	"github.com/elastic/elastic-agent/testing/upgradetest"
+)
+
+func TestStandaloneDowngradeToPreviousSnapshotBuild(t *testing.T) {
+	define.Require(t, define.Requirements{
+		Local: false, // requires Agent installation
+		Sudo:  true,  // requires Agent installation
+	})
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// Start at the build version as we want to test the retry
+	// logic that is in the build.
+	startFixture, err := define.NewFixture(t, define.Version())
+	require.NoError(t, err)
+
+	// Upgrade to an old build, see `BackwardTwoMinors` for why.
+	upgradeToVersion, err := upgradetest.BackwardTwoMinors(define.Version())
+	require.NoError(t, err)
+	endFixture, err := atesting.NewFixture(
+		t,
+		upgradeToVersion,
+		atesting.WithFetcher(atesting.ArtifactFetcher()),
+	)
+	require.NoError(t, err)
+
+	t.Logf("Testing Elastic Agent upgrade from %s to %s...", define.Version(), upgradeToVersion)
+
+	err = upgradetest.PerformUpgrade(ctx, startFixture, endFixture, t)
+	assert.NoError(t, err)
+}
```
This tests was testing the ability to upgrade/downgrade to a specific snapshot build before the refactor... I don't think it's doing that now...

blakerouse replied:
Sadly the title of the test did not make that clear. Most test where so hard to follow and read I could not fully understand what the test was doing.

If you can provide me some more detail on what this test is really doing, I can fix it.

pchila replied:
The original test would select the second-to-last (penultimate) snapshot build from Artifact API, use the buildid to issue a `elastic-agent upgrade x.y.z-SNAPSHOT+<buildid>` in order to distinguish the result from a regular `elastic-agent upgrade x.y.z-SNAPSHOT`

It would then proceed to verify that the commit of the upgraded agent would correspond to the specific build selected, proving that the feature works.

There was also an ugly workaround because the commit hash reported by Artifact API is not always the correct one, but with your approach of creating an endFixture we may sidestep that one if the artifactFetcher is able to pinpoint a specific snapshot build.

blakerouse replied:
Okay I will go back and look at the original test and see if there is a clean and simple way of finding the `<buildid>`. The artifact fetcher in the testing library might need to be updated to use that `+<buildid>` as well to pull the correct build.

blakerouse replied:
I have updated the test so it is testing what used to be tested. Sorry about that.

I had to add support for the testing framework to work with build ids in the version name as well.

pchila commented on testing/upgradetest/upgrader.go:
```diff
@@ -0,0 +1,358 @@
+	if upgradeOpts.sourceURI != "" {
+		// specific ---source-uri
+		upgradeCmdArgs = append(upgradeCmdArgs, "--source-uri", upgradeOpts.sourceURI)
+	} else if upgradeOpts.customPgp == nil {
```
Not sure why this is an `else if` clause instead of being a simple `if`... 
Are we saying that this function will ignore PGP options if a custom `source-uri` is provided ? 
Shouldn't that decision be left to the test calling this function with the appropriate options?

blakerouse replied:
True. I will update.

blakerouse replied:
Updated.

pchila reviewed (changes requested):
There is at least 1 test that has changed meaning in the refactor: ` TestStandaloneDowngradeToPreviousSnapshotBuild` is supposed to test the upgrade in form of `elastic-agent upgrade x.y.z-SNAPSHOT+<buildid>` and from what I see it doesn't do that anymore but it just performs a downgrade 2 minors back

There are a few other comments on other parts of the code but another element that pops is that we use `version.GetPreviousMinor()` more even if we know that it has some limitation as it tries to guess what the previous minor would be.
Maybe you want to take a look at https://github.com/elastic/elastic-agent/pull/3458 (which I imagine is now replaced by this) for a possible alternative of selecting previous versions

blakerouse commented:
The modreplace has been removed but the lint is still failing, not clear as to why.

pchila commented on testing/upgradetest/upgrader.go:
```diff
@@ -0,0 +1,365 @@
+
+	logger.Logf("Upgrading from version %q to version %q", startParsedVersion, endVersionInfo.Binary.String())
+
+	upgradeCmdArgs := []string{"upgrade", endVersionInfo.Binary.String()}
+	if upgradeOpts.customPgp == nil {
+		// unless a custom PGP configuration is provided, upgrade is always using --source-uri
+		if upgradeOpts.sourceURI != "" {
+			// specific ---source-uri
+			upgradeCmdArgs = append(upgradeCmdArgs, "--source-uri", upgradeOpts.sourceURI)
+		} else {
+			// --source-uri from the endFixture
+			srcPkg, err := endFixture.SrcPackage(ctx)
+			if err != nil {
+				return fmt.Errorf("failed to get end agent source package path: %w", err)
+			}
+			sourceURI := "file://" + filepath.Dir(srcPkg)
+			upgradeCmdArgs = append(upgradeCmdArgs, "--source-uri", sourceURI)
+		}
+	} else {
+		if upgradeOpts.sourceURI != "" {
+			upgradeCmdArgs = append(upgradeCmdArgs, "--source-uri", upgradeOpts.sourceURI)
+		}
+
+		if len(upgradeOpts.customPgp.PGP) > 0 {
+			upgradeCmdArgs = append(upgradeCmdArgs, "--pgp", upgradeOpts.customPgp.PGP)
+		}
+
+		if len(upgradeOpts.customPgp.PGPUri) > 0 {
+			upgradeCmdArgs = append(upgradeCmdArgs, "--pgp-uri", upgradeOpts.customPgp.PGPUri)
+		}
+
+		if len(upgradeOpts.customPgp.PGPPath) > 0 {
+			upgradeCmdArgs = append(upgradeCmdArgs, "--pgp-path", upgradeOpts.customPgp.PGPPath)
+		}
+	}
```
This still doesn't look great (I know it comes from the very specific logic of upgrade tests with PGP).
I would be tempted to separate the customPGP != nil and customPGP == nil at cost of a little duplication for the sourceURI logic but I leave that choice up to you

blakerouse replied:
I will have to go with a `*string` for the `sourceURI` to use that logic. That means the GPG tests would need to set `WithSourceURI("")` that way it doesn't use the local `sourceURI`. I can change to that way of operation.

blakerouse replied:
I changed to to how I described. See if you like that way better.

pchila replied:
Looks good now! Thanks! :+1:

pchila commented on testing/upgradetest/upgrader.go:
```diff
@@ -190,25 +190,20 @@ func PerformUpgrade(
+		upgradeCmdArgs = append(upgradeCmdArgs, "--source-uri", *upgradeOpts.sourceURI)
+	}
 
+	if upgradeOpts.customPgp != nil {
```
Much easier to follow (at least for me). Thank you :+1:

pchila commented:
/test

ycombinator commented:
The conflicts are due to me merging #3085, introducing a new upgrade integration test, `TestStandaloneUpgradeFailsRestart`. @blake.rouse I'll make a PR to thiw refactoring PR fixing up the test I just added to bring it in line with your improvements.  I just wanted to get #3085 in since it's been open for two months!

blakerouse commented:
@ycombinator Don't worry about it. I have fixed the conflict and updated the test to use the new patterns.

blakerouse commented:
@pchila I believe this is ready for a final review. I have fixed all conflicts and issues you pointed out.

pchila reviewed (changes requested):
The  `TestStandaloneDowngradeToSpecificSnapshotBuild` looks better but we need to select the penultimate build to have a result that is noticeably different from a normal upgrade to a snapshot version

blakerouse commented on testing/integration/upgrade_downgrade_test.go:
```diff
@@ -8,41 +8,74 @@ package integration
+	// get all the builds of the snapshot version (need to pass x.y.z-SNAPSHOT format)
+	builds, err := aac.GetBuildsForVersion(ctx, latestSnapshotVersion.VersionWithPrerelease())
+	require.NoError(t, err)
+	if len(builds.Builds) == 0 {
+		t.Skipf("no SNAPSHOT's exists for version %s", latestSnapshotVersion.VersionWithPrerelease())
+	}
```
I never understood why the test needed 2 builds. I get it now. Next time more comments on why things are done the way they where in the test would be helpful. I just keep looking at it thinking why does it need 2? Its only upgrading to 1 build.

blakerouse replied:
Done.

blakerouse commented on testing/integration/upgrade_downgrade_test.go:
```diff
@@ -8,41 +8,74 @@ package integration
+	}
+
+	// find the specific build to use for the test
+	upgradeVersionString := builds.Builds[0]
```
Done.

blakerouse commented on pkg/testing/fetcher_artifact.go:
```diff
@@ -168,13 +170,35 @@ func findURI(ctx context.Context, doer httpDoer, version string) (string, error)
 	return "", fmt.Errorf("uri not detected")
 }
 
+func splitBuildID(version string) (string, string) {
+	split := strings.SplitN(version, "+", 2)
+	if len(split) == 1 {
+		// no build ID
+		return split[0], ""
+	}
+	return split[0], split[1]
+}
+
+func stripSnapshot(version string) string {
+	if strings.HasSuffix(version, "-SNAPSHOT") {
+		return strings.TrimSuffix(version, "-SNAPSHOT")
+	}
+	return version
+}
+
```
Its simple and this module doesn't import that module at the moment, so I am just going to keep it here.

blakerouse commented:
@pchila I have fixed that test and fixed the conflicts.

@ycombinator I have fixed the two tests in conflict if you want to give them a look over. Both where new tests that you added.

ycombinator commented on testing/integration/upgrade_standalone_inprogress.go:
```diff
@@ -0,0 +1,97 @@
+	// before trying to perform another upgrade
+	err = upgradetest.WaitHealthyAndVersion(ctx, startFixture, endVersionInfo.Binary, 2*time.Minute, 10*time.Second, t)
+	if err != nil {
+		// agent never got healthy, but we need to ensure the watcher is stopped before continuing
```
It wasn't immediately obvious to me _why_ the watcher was being killed here.  I think it's so it doesn't interfere with _other_ tests (as opposed to having anything to do with continuing on in this test, since this test should now fail due to the `require.NoError(t, err)` on line 79), right?  Could you clarify that in the code comment here please?

blakerouse replied:
I added more to the comment to make this clear. The reason you gave is the correct.

ycombinator commented on testing/integration/upgrade_standalone_inprogress.go:
```diff
@@ -0,0 +1,97 @@
+	require.NotNil(t, err)
+	require.Contains(t, string(upgradeOutput), "an upgrade is already in progress; please try again later.")
+
+	// killTimeout is greater than timeout as the watcher should have been
+	// stopped on its own, and we don't want this test to hide that fact
```
👍

ycombinator commented on testing/integration/upgrade_rollback_test.go:
```diff
@@ -0,0 +1,195 @@
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// Start at the build version as we want to test the retry
+	// logic that is in the build.
```
I think this comment is not relevant to this test?

blakerouse replied:
It is relevant.

ycombinator commented on testing/integration/upgrade_rollback_test.go:
```diff
@@ -0,0 +1,195 @@
+	// wait for the agent to be healthy and back at the start version
+	err = upgradetest.WaitHealthyAndVersion(ctx, startFixture, startVersionInfo.Binary, 10*time.Minute, 10*time.Second, t)
+	if err != nil {
+		// agent never got healthy, but we need to ensure the watcher is stopped before continuing
+		// this kills the watcher instantly and waits for it to be gone before continuing
```
Same comment as https://github.com/elastic/elastic-agent/pull/3477/files#r1341596538 about the "continuing" being unclear.

ycombinator commented:
> @ycombinator I have fixed the two tests in conflict if you want to give them a look over. Both where new tests that you added.

@blakerouse I reviewed the two tests. The logic in both looks good and even improved towards the end of the tests! I left some feedback about clarifying comments, that's all.

pchila commented on pkg/testing/fetcher_artifact.go:
```diff
@@ -68,6 +68,7 @@ func (f *artifactFetcher) Fetch(ctx context.Context, operatingSystem string, arc
 	preVersion := version
+	version, _ = splitBuildID(version)
 	if uri == "" {
 		if !strings.HasSuffix(version, "-SNAPSHOT") {
```
If we have a snapshot with a buildId we need to prepare the uri in the format of x.y.z-<buildid>, for example
`https://snapshots.elastic.co/8.11.0-2574a657/downloads/beats/elastic-agent/elastic-agent-8.11.0-SNAPSHOT-linux-x86_64.tar.gz` (we can see the build details from https://artifacts-api.elastic.co/v1/versions/8.11.0-SNAPSHOT/builds/8.11.0-2574a657 )

blakerouse commented:
@pchila You are correct it was not, my most recent commit fixed it. See here - https://github.com/elastic/elastic-agent/pull/3477/commits/967969f1d2e640b923160d5e666d690821c43871
